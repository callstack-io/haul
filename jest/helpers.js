/**
 * Copyright 2017-present, Callstack.
 * All rights reserved.
 *
 * @flow
 */

'use strict';

const traverse = require('traverse'); // eslint-disable-line
const path = require('path');
const excapeStringRegex = require('escape-string-regexp');

const flushPromises = (): Promise<any> =>
  new Promise(resolve => setTimeout(resolve));

const replacePathsInObject = (object: mixed) => {
  return traverse(object).map(
    entry =>
      typeof entry === 'string'
        ? entry
            .replace(/\/.*\/node_modules/, '<<NODE_MODULE>>')
            .replace(
              new RegExp(
                `^${excapeStringRegex(path.resolve(__dirname, '..'))}`
              ),
              '<<REPLACED>>'
            )
        : entry
  );
};

module.exports = {
  flushPromises,
  replacePathsInObject,
};
