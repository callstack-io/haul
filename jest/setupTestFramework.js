/* global jest */

jest.mock('chalk');

// Integration tests are quite slow to run on CI, feel free to adjust as needed
jest.setTimeout(80000);
