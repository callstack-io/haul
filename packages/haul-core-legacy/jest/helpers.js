/**
 * Copyright 2017-present, Callstack.
 * All rights reserved.
 *
 * @flow
 */

'use strict';

const traverse = require('traverse'); // eslint-disable-line
const path = require('path');
const excapeStringRegex = require('escape-string-regexp');

const flushPromises = (): Promise<any> =>
  new Promise(resolve => setTimeout(resolve));

const replacePathsInObject = (object: mixed) => {
  return traverse(object).map(
    entry =>
      typeof entry === 'string'
        ? entry
            .replace(
              new RegExp(`^${excapeStringRegex(path.resolve('/'))}`),
              '/'
            )
            .replace(new RegExp('\\\\', 'g'), '/')
            .replace(/\/.*\/node_modules/, '<<NODE_MODULE>>')
            .replace(
              new RegExp(
                process.platform === 'win32'
                  ? `^${excapeStringRegex(
                      path
                        .resolve(__dirname, '..')
                        .slice(2)
                        .replace(new RegExp('\\\\', 'g'), '/')
                    )}`
                  : `^${excapeStringRegex(path.resolve(__dirname, '..'))}`
              ),
              '<<REPLACED>>'
            )
        : entry
  );
};

const fixWorkers = (object: mixed) => {
  return traverse(object).map(entry => {
    if (typeof entry === 'object' && 'workers' in entry) {
      return Object.assign({}, entry, { workers: '<<CPUS>>' });
    }

    return entry;
  });
};

module.exports = {
  flushPromises,
  replacePathsInObject,
  fixWorkers,
};
