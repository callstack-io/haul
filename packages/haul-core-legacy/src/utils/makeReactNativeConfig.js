/**
 * Copyright 2017-present, Callstack.
 * All rights reserved.
 *
 * @flow
 */
/* eslint-disable no-param-reassign */

const path = require('path');
const getPolyfills = require('./getPolyfills');

type WebpackEntry = string | Array<string> | Object;

function injectPolyfillIntoEntry({
  entry: userEntry,
  root,
  initializeCoreLocation = 'node_modules/react-native/Libraries/Core/InitializeCore.js',
  dev = true,
  disableHotReloading = false,
}: {
  entry: WebpackEntry,
  root: string,
  initializeCoreLocation?: string,
  dev?: boolean,
  disableHotReloading?: boolean,
}) {
  const reactNativeHaulEntries = [
    ...getPolyfills(),
    require.resolve(path.join(root, initializeCoreLocation)),
  ];

  if (dev && !disableHotReloading) {
    reactNativeHaulEntries.push(require.resolve('../../hot/patch.js'));
  }

  return makeWebpackEntry(userEntry, reactNativeHaulEntries);
}

function makeWebpackEntry(
  userEntry: WebpackEntry,
  otherEntries: Array<string>
): WebpackEntry {
  if (typeof userEntry === 'string') {
    return [...otherEntries, userEntry];
  }
  if (Array.isArray(userEntry)) {
    return [...otherEntries, ...userEntry];
  }
  if (typeof userEntry === 'object') {
    const chunkNames = Object.keys(userEntry);
    return chunkNames.reduce((entryObj: Object, name: string) => {
      // $FlowFixMe
      const chunk = userEntry[name];
      if (typeof chunk === 'string') {
        entryObj[name] = [...otherEntries, chunk];
        return entryObj;
      } else if (Array.isArray(chunk)) {
        entryObj[name] = [...otherEntries, ...chunk];
        return entryObj;
      }
      return chunk;
    }, {});
  }
  return userEntry;
}

module.exports = {
  injectPolyfillIntoEntry,
};
