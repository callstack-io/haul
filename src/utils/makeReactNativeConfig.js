/**
 * Copyright 2017-present, Callstack.
 * All rights reserved.
 *
 * makeReactNativeConfig.js
 *
 * @flow
 */
const webpack = require('webpack');
const HappyPack = require('happypack');
const path = require('path');
const findProvidesModule = require('./findProvidesModule');

const PLATFORMS = ['ios', 'android'];

type ConfigOptions = {
  port: number,
  platform: 'ios' | 'android',
  cwd: string,
  dev: boolean
};

// @todo type this
type WebpackConfig = {
  entry: Array<string>
};

type WebpackConfigFactory =
  | ((ConfigOptions, WebpackConfig) => WebpackConfig)
  | WebpackConfig;

/**
 * Returns default config based on environment
 */
const getDefaultConfig = ({ platform, cwd, dev, port }): WebpackConfig => ({
  // Default polyfills and entry-point setup
  entry: [require.resolve('./polyfillEnvironment.js')],
  // Built-in loaders
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'happypack/loader?id=babel',
        exclude: /node_modules\/(?!react)/
      },
      {
        test: /\.(bmp|gif|jpg|jpeg|png)$/,
        loader: require.resolve('../loaders/asset-loader')
      },
      {
        test: /\.json$/,
        loader: 'json-loader'
      }
    ]
  },
  output: {
    path: '/',
    filename: `index.${platform}.bundle`
  },
  // Default plugins
  plugins: [
    new webpack.DefinePlugin({
      __DEV__: dev
    }),
    // Use HappyPack to speed up Babel build times
    // significantly
    new HappyPack({
      id: 'babel',
      loaders: [
        `babel-loader?presets[]=react-native&plugins[]=${require.resolve('./fixRequireIssues')}`
      ],
      verbose: false
    })
  ],
  // Default resolve
  resolve: {
    alias: findProvidesModule([path.resolve(cwd, 'node_modules/react-native')]),
    extensions: [`.${platform}.js`, '.js']
  }
});

/**
 * Creates an array of configs based on changing `env` for every
 * platform and returns
 */
function makeReactNativeConfig(
  userWebpackConfig: WebpackConfigFactory,
  options: ConfigOptions
): WebpackConfig {
  const configs = PLATFORMS.map(platform => {
    const env = Object.assign({}, options, { platform });
    const defaultWebpackConfig = getDefaultConfig(env);

    const config = Object.assign(
      {},
      defaultWebpackConfig,
      typeof userWebpackConfig === 'function'
        ? userWebpackConfig(env, defaultWebpackConfig)
        : userWebpackConfig
    );

    // For simplicity, we don't require users to extend
    // default config.entry but do it for them.
    config.entry = defaultWebpackConfig.entry.concat(config.entry);

    return config;
  });

  return configs[PLATFORMS.indexOf(options.platform)];
}

module.exports = makeReactNativeConfig;
