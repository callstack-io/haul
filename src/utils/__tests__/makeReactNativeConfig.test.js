/**
 * Copyright 2017-present, Callstack.
 * All rights reserved.
 *
 * @flow
 */

import path from 'path';
import { replacePathsInObject } from 'jest/helpers'; // eslint-disable-line import/no-unresolved
import makeReactNativeConfig from '../makeReactNativeConfig';

test('creates config from defaults', () => {
  const webpackConfig = require('./fixtures/webpack.config.js');
  const [configs, platforms] = makeReactNativeConfig(webpackConfig, {
    dev: true,
    root: path.resolve(__dirname, 'fixtures'),
  });

  expect(replacePathsInObject(configs)).toMatchSnapshot(
    'creates config from defaults (configs)'
  );
  expect(platforms).toMatchSnapshot('creates config from defaults (platforms)');
});

test('merges existing config', () => {
  const webpackConfig = require('./fixtures/webpack.custom.config.js');
  const [configs] = makeReactNativeConfig(webpackConfig, {
    dev: true,
    root: path.resolve(__dirname, 'fixtures'),
  });

  expect(replacePathsInObject(configs)).toMatchSnapshot(
    'merges existing config'
  );
});
