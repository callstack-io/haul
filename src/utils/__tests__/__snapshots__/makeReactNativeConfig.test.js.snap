// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeReactNativeConfig creates config from defaults: diff ios/android config 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -45,18 +45,18 @@
          \\"test\\": /\\\\.(aac|aiff|bmp|caf|gif|html|jpeg|jpg|m4a|m4v|mov|mp3|mp4|mpeg|mpg|obj|otf|pdf|png|psd|svg|ttf|wav|webm|webp)$/,
          \\"use\\": Object {
            \\"loader\\": \\"<<REPLACED>>/src/loaders/assetLoader.js\\",
            \\"query\\": Object {
              \\"bundle\\": undefined,
-             \\"platform\\": \\"ios\\",
+             \\"platform\\": \\"android\\",
              \\"root\\": \\"<<REPLACED>>/src/utils/__tests__/fixtures\\",
            },
          },
        },
      ],
    },
-   \\"name\\": \\"ios\\",
+   \\"name\\": \\"android\\",
    \\"optimization\\": Object {
      \\"concatenateModules\\": true,
      \\"minimize\\": false,
      \\"minimizer\\": Array [
        UglifyJsPlugin {
@@ -81,11 +81,11 @@
        },
      ],
      \\"namedModules\\": true,
    },
    \\"output\\": Object {
-     \\"filename\\": \\"index.ios.bundle\\",
+     \\"filename\\": \\"index.android.bundle\\",
      \\"path\\": \\"<<REPLACED>>/src/utils/__tests__/fixtures\\",
      \\"publicPath\\": \\"http://localhost:8081/\\",
    },
    \\"plugins\\": Array [
      CaseSensitivePathsPlugin {
@@ -149,11 +149,11 @@
    \\"resolve\\": Object {
      \\"alias\\": Object {
        \\"react-proxy\\": \\"@zamotany/react-proxy\\",
      },
      \\"extensions\\": Array [
-       \\".ios.js\\",
+       \\".android.js\\",
        \\".native.js\\",
        \\".js\\",
      ],
      \\"mainFields\\": Array [
        \\"react-native\\","
`;
